### ISSUE TITLE
Indentation: lambda don't respect lineWrappingIndentation when forceStrictCondition is false

### ISSUE THREAD

**Issue by aruberto** [checkstyle/checkstyle#3342]
/var/tmp $ javac Test.java

```
```

/var/tmp $ cat Test.java

```
package com.github.aruberto;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public class Test {

    public static void someFunction(SomeUtilClass util) {

        // This fails indentation rule
        util.myLambdaUtil("FIRST_ARG",
                          (string) -> System.out.println(string.trim()), //line 13
                          "SECOND_ARG",
                          () -> "WHAT WHAT!");      //line 15

        // This passes indentation rule
        util.myLambdaUtil("FIRST_ARG",
            (string) -> System.out.println(string.trim()),
                          "SECOND_ARG",
            () -> "WHAT WHAT!");

        // This fails indentation rule
        Function<String, String> someFunction1 =
                (string) -> {                       // line 25
                    if (string.contains("abc")) {
                        return "SWEET!";
                    } else if (string.contains("123")) {
                        return "COOL!";
                    } else {
                        return "BOO!";
                    }
                };                                 // line 33

        // This passes indentation rule
        Function<String, String> someFunction2 =
        (string) -> {
            if (string.contains("abc")) {
                return "SWEET!";
            } else if (string.contains("123")) {
                return "COOL!";
            } else {
                return "BOO!";
            }
        };
    }

    interface SomeUtilClass {

        void myLambdaUtil(String firstArg,
                          Consumer<String> firstLambda,
                          String secondArg,
                          Supplier<String> secondLambda);
    }
}    
```

/var/tmp $ cat config.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
    "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="charset" value="UTF-8"/>

  <property name="severity" value="warning"/>

  <property name="fileExtensions" value="java, properties, xml"/>

  <module name="FileTabCharacter">
    <property name="eachLine" value="true"/>
  </module>

  <module name="TreeWalker">
    <module name="Indentation">
      <property name="basicOffset" value="4"/>
      <property name="braceAdjustment" value="0"/>
      <property name="caseIndent" value="4"/>
      <property name="throwsIndent" value="8"/>
      <property name="arrayInitIndent" value="4"/>
      <property name="lineWrappingIndentation" value="0"/>
      <property name="forceStrictCondition" value="false"/>
    </module>
  </module>
</module>                                                                         
```

/var/tmp $ java -jar checkstyle-7.0-all.jar -c config.xml Test.java

```
Starting audit...                                                                                                                                                                               
[WARN] Test.java:13: 'lambda arguments' have incorrect indentation level 26, expected level should be one of the following: 8, 12. [Indentation] 
[WARN] Test.java:15: 'lambda arguments' have incorrect indentation level 26, expected level should be one of the following: 8, 12. [Indentation] 
[WARN] Test.java:25: 'lambda arguments' have incorrect indentation level 16, expected level should be 8. [Indentation]                           
[WARN] Test.java:26: 'if' have incorrect indentation level 20, expected level should be 12. [Indentation]                                        
[WARN] Test.java:27: 'if' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                  
[WARN] Test.java:28: 'if rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                                 
[WARN] Test.java:29: 'if' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                  
[WARN] Test.java:30: 'if rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                                 
[WARN] Test.java:31: 'else' child have incorrect indentation level 24, expected level should be 16. [Indentation]                                
[WARN] Test.java:32: 'else rcurly' have incorrect indentation level 20, expected level should be 12. [Indentation]                               
[WARN] Test.java:33: 'block rcurly' have incorrect indentation level 16, expected level should be 8. [Indentation]                               
Audit done.                                                                                                                                                                                     
```

---

Describe what you expect in details.

Given that forceStrictCondition is false and lineWrappingIndentation = 0, I should be able to indent lambda expressions on wrapping lines at any indent level > 0. 

In First example, the String arg on line 14 is aligned and passes indentation rule (indention of 26) but the Consumer and Supplier args on line 13 and 15 are forced to indent at 8 or 12.

In second example, I define my Function on a new line after assignment operator (lines 25-33). This is indented by Intellij to 16 as it respects current line wrapping settings. But checkstyle want an indentation of 8 for wrapping line and all subsequent lines of lambda expression.
##

**Comment by mld-ger** [comment]
Is there an update on this issue? I have the same problem when doing:
```java
list.stream()
  .map(element -> new MappedElement(element)
  .collect(toList());
```

Config is:
```xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name = "Checker">
    <property name="charset" value="UTF-8"/>

    <property name="severity" value="warning"/>

    <property name="fileExtensions" value="java, properties, xml"/>

    <module name="TreeWalker">
        <property name="tabWidth" value="4"/>
        <module name="RegexpSinglelineJava">
             <property name="format" value="^\t* "/>
             <property name="message" value="Indent must use tab characters"/>
             <property name="ignoreComments" value="true"/>
       </module>
        <module name="OuterTypeFilename"/>
        <module name="IllegalTokenText">
            <property name="tokens" value="STRING_LITERAL, CHAR_LITERAL"/>
            <property name="format" value="\\u00(08|09|0(a|A)|0(c|C)|0(d|D)|22|27|5(C|c))|\\(0(10|11|12|14|15|42|47)|134)"/>
            <property name="message" value="Avoid using corresponding octal or Unicode escape."/>
        </module>
        <module name="AvoidEscapedUnicodeCharacters">
            <property name="allowEscapesForControlCharacters" value="true"/>
            <property name="allowByTailComment" value="true"/>
            <property name="allowNonPrintableEscapes" value="true"/>
        </module>
        <module name="LineLength">
            <property name="max" value="120"/>
            <property name="ignorePattern" value="^package.*|^import.*|a href|href|http://|https://|ftp://"/>
        </module>
        <module name="AvoidStarImport"/>
        <module name="OneTopLevelClass"/>
        <module name="NoLineWrap"/>
        <module name="EmptyBlock">
            <property name="option" value="TEXT"/>
            <property name="tokens" value="LITERAL_TRY, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE, LITERAL_SWITCH"/>
        </module>
        <module name="NeedBraces"/>
        <module name="LeftCurly">
            <property name="maxLineLength" value="120"/>
        </module>
        <module name="RightCurly"/>
        <module name="RightCurly">
            <property name="option" value="alone"/>
            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, LITERAL_DO, STATIC_INIT, INSTANCE_INIT"/>
        </module>
        <module name="WhitespaceAround">
            <property name="allowEmptyConstructors" value="true"/>
            <property name="allowEmptyMethods" value="true"/>
            <property name="allowEmptyTypes" value="true"/>
            <property name="allowEmptyLoops" value="true"/>
            <message key="ws.notFollowed"
             value="WhitespaceAround: ''{0}'' is not followed by whitespace. Empty blocks may only be represented as '{}' when not part of a multi-block statement (4.1.3)"/>
             <message key="ws.notPreceded"
             value="WhitespaceAround: ''{0}'' is not preceded with whitespace."/>
        </module>
        <module name="OneStatementPerLine"/>
        <module name="MultipleVariableDeclarations"/>
        <module name="ArrayTypeStyle"/>
        <module name="MissingSwitchDefault"/>
        <module name="FallThrough"/>
        <module name="UpperEll"/>
        <module name="ModifierOrder"/>
        <module name="EmptyLineSeparator">
            <property name="allowNoEmptyLineBetweenFields" value="true"/>
        </module>
        <module name="SeparatorWrap">
            <property name="tokens" value="DOT"/>
            <property name="option" value="nl"/>
        </module>
        <module name="SeparatorWrap">
            <property name="tokens" value="COMMA"/>
            <property name="option" value="EOL"/>
        </module>
        <module name="PackageName">
            <property name="format" value="^[a-z]+(\.[a-z][a-z0-9]*)*$"/>
            <message key="name.invalidPattern"
             value="Package name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="TypeName">
            <message key="name.invalidPattern"
             value="Type name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="MemberName">
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <message key="name.invalidPattern"
             value="Member name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="ParameterName">
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <message key="name.invalidPattern"
             value="Parameter name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="LocalVariableName">
            <property name="tokens" value="VARIABLE_DEF"/>
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9]*$"/>
            <property name="allowOneCharVarInForLoop" value="true"/>
            <message key="name.invalidPattern"
             value="Local variable name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="ClassTypeParameterName">
            <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
            <message key="name.invalidPattern"
             value="Class type name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="MethodTypeParameterName">
            <property name="format" value="(^[A-Z][0-9]?)$|([A-Z][a-zA-Z0-9]*[T]$)"/>
            <message key="name.invalidPattern"
             value="Method type name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="NoFinalizer"/>
        <module name="GenericWhitespace">
            <message key="ws.followed"
             value="GenericWhitespace ''{0}'' is followed by whitespace."/>
             <message key="ws.preceded"
             value="GenericWhitespace ''{0}'' is preceded with whitespace."/>
             <message key="ws.illegalFollow"
             value="GenericWhitespace ''{0}'' should followed by whitespace."/>
             <message key="ws.notPreceded"
             value="GenericWhitespace ''{0}'' is not preceded with whitespace."/>
        </module>
        <module name="Indentation">
            <property name="basicOffset" value="4"/>
            <property name="braceAdjustment" value="0"/>
            <property name="caseIndent" value="4"/>
            <property name="throwsIndent" value="4"/>
            <property name="lineWrappingIndentation" value="8"/>
            <property name="arrayInitIndent" value="4"/>
        </module>
        <module name="OverloadMethodsDeclarationOrder"/>
        <module name="VariableDeclarationUsageDistance"/>
        <module name="MethodParamPad"/>
        <module name="OperatorWrap">
            <property name="option" value="NL"/>
            <property name="tokens" value="BAND, BOR, BSR, BXOR, DIV, EQUAL, GE, GT, LAND, LE, LITERAL_INSTANCEOF, LOR, LT, MINUS, MOD, NOT_EQUAL, PLUS, QUESTION, SL, SR, STAR "/>
        </module>
        <module name="AnnotationLocation">
            <property name="tokens" value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF"/>
        </module>
        <module name="AnnotationLocation">
            <property name="tokens" value="VARIABLE_DEF"/>
            <property name="allowSamelineMultipleAnnotations" value="true"/>
        </module>
        <module name="NonEmptyAtclauseDescription"/>
        <module name="JavadocTagContinuationIndentation"/>
        <module name="SummaryJavadocCheck">
            <property name="forbiddenSummaryFragments" value="^@return the *|^This method returns |^A [{]@code [a-zA-Z0-9]+[}]( is a )"/>
        </module>
        <module name="JavadocParagraph"/>
        <module name="AtclauseOrder">
            <property name="tagOrder" value="@param, @return, @throws, @deprecated"/>
            <property name="target" value="CLASS_DEF, INTERFACE_DEF, ENUM_DEF, METHOD_DEF, CTOR_DEF, VARIABLE_DEF"/>
        </module>
        <module name="JavadocMethod">
            <property name="scope" value="public"/>
            <property name="allowMissingParamTags" value="true"/>
            <property name="allowMissingThrowsTags" value="true"/>
            <property name="allowMissingReturnTag" value="true"/>
            <property name="minLineCount" value="2"/>
            <property name="allowedAnnotations" value="Override, Test"/>
            <property name="allowThrowsTagsForSubclasses" value="true"/>
        </module>
        <module name="MethodName">
            <property name="format" value="^[a-z][a-z0-9][a-zA-Z0-9_]*$"/>
            <message key="name.invalidPattern"
             value="Method name ''{0}'' must match pattern ''{1}''."/>
        </module>
        <module name="SingleLineJavadoc">
            <property name="ignoreInlineTags" value="false"/>
        </module>
        <module name="EmptyCatchBlock">
            <property name="exceptionVariableName" value="expected"/>
        </module>
        <module name="CommentsIndentation"/>
    </module>
</module>
```

**Comment by rnveach** [comment]
> Is there an update on this issue?

No. You would see something here if there was.

**Comment by romani** [comment]
@mld-ger , @aruberto, it is unlikely that we will address this issue in next 2-3 month, if you have time - you always welcome with PullRequest.

**Comment by beihaifeiwu** [comment]
+1

**Comment by kdubb** [comment]
+1

As Java 8 has been mainstream for quite a while at this point. This bug makes the `Indentation` check completely unusable in the age of lambdas.

**Comment by Jonarzz** [comment]
+1

**Comment by jakub-bochenski** [comment]
Would it be possible to at least add an option to disable the malfunctioning (lambda-related) part? Right now we were forced to disable the whole check

**Comment by romani** [comment]
@jakub-bochenski , thanks for making a post to bring our attention.

Unfortunately we have lack of time to look at it. We need smb from outside to make a contribution by fixing this issue. If smb know java , it is not that over complicated, whole logic is at https://github.com/checkstyle/checkstyle/tree/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation , development is easy to start http://checkstyle.sourceforge.net/beginning_development.html.

I placed some bounty on this issue from donations budget:
https://www.bountysource.com/issues/35888832-indentation-lambda-don-t-respect-linewrappingindentation-when-forcestrictcondition-is-false
I hope this will let somebody in web to spend several hours on this issue and have a beer after that.

If you need this issue to be fixed, please feel free to add more incentive for contributor. Just to let him have beer for himself and friends while watching WorldCup .

**Comment by sdoeringNew** [comment]
After seeing the code I think the problem goes deeper.

Almost no indentation handler respects the ``forceStrictCondition`` flag - only ``LineWrappingHandler`` and the ``TryHandler``.
Adding this additional check to the ``LambdaHandler`` is _relatively_ easy.

But the ``BlockParentHandler`` has issues, too. And that's the real headache here as it seems to be the most complex class in the indentation package.

---

Instead of fixing every class in this package would it be possible to refactor just the ``IndentLevel#isAcceptable(int)`` method?
```Java
    private final boolean forceStrictCondition;

    public boolean isAcceptable(int indent) {
        if(forceStrictCondition) {
            return levels.get(indent);
        } else {
            // no strict condition - accept all indentations if they are
            // at least higher or equal the minimum accepted indentation
            return indent >= levels.nextSetBit(0);
        }
    }
```

**Comment by binaryoverload** [comment]
3 years later and 4 java versions later, there's still no fix for this issue?

**Comment by romani** [comment]
@binaryoverload , you or anybody else are welcome to contribute.

**Comment by Fernal73** [comment]
I was wondering if you could allow multiple indentation values instead of single ones? Is that heretical?

**Comment by Fernal73** [comment]
https://www.google.com/url?q=https://github.com/google/google-java-format/issues/19?email_source%3Dnotifications%26email_token%3DAF4R3PKWMTL4J3E4FBZLFH3P2ZXN5A5CNFSM4BW4IBQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXZRE2I%23issuecomment-502469225&source=gmail&ust=1560792141672000&usg=AFQjCNHNTqe7bNvRvtLdOt_8wTDtFphT9Q

**Comment by romani** [comment]
unlikely .... we cannot even maintain Check in current state , too much bugs in it, so making it even more complicated it not doable. We can come back to this idea after we [fix all existing problems in this Check](https://github.com/checkstyle/checkstyle/issues?q=is%3Aopen+is%3Aissue+label%3Aindentaion).

**Comment by Fernal73** [comment]
I'm just pointing out that disabling the check and letting Google Java
Formatter do its job is probably the best option.
The other option might be to use a formatter that uses Google Java style
and configure it to indent fewer spaces for line wrapping. I have not
looked at your code as yet but from the comments it's evident the code base
is fragile.

Personally, I think a range might be a, better option for values to be
used. That will, however, increase the complexity of the coded solution and
that's the last thing you need. It is not a comment on the efficacy of such
a rule. It would certainly make it from flexible and rekaxable.

**Comment by Fernal73** [comment]
Relaxable. Auto correct is on the blink!

**Comment by Fernal73** [comment]
Rekaxable.

**Comment by Abhishek-kumar09** [comment]
@romani @rnveach
For confirmation:
```java
 util.myLambdaUtil("FIRST_ARG",
                  (string) -> System.out.println(string.trim()), //line 13
                  "SECOND_ARG",
                  () -> "WHAT WHAT!"); 
```
the code above is not formatted correctly, It should be something like this:  
```java
        util.myLambdaUtil("FIRST_ARG",
            (string) -> System.out.println(string.trim()), //line 13
            "SECOND_ARG",
            () -> "WHAT WHAT!"); 
```
1. But since force line conndition is `false`, should we allow that first formatting?

> Given that forceStrictCondition is false and lineWrappingIndentation = 0, I should be able to indent lambda expressions on wrapping lines at any indent level > 0.

2. Should it be allowed by checkstyle?

**Comment by romani** [comment]
@Abhishek-kumar09, thanks a lot for your hard work on this
